// locize/dist/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var locizePlugin = require('./locizePlugin.js');
var startStandalone = require('./startStandalone.js');
require('./api/handleEditKey.js');
require('./api/handleCommitKey.js');
require('./api/handleCommitKeys.js');
require('./api/handleConfirmInitialized.js');
require('./api/handleRequestPopupChanges.js');
require('./api/handleRequestResourceBundle.js');
require('./api/handleSelectedKeys.js');
require('./api/handleIsLocizeEnabled.js');
require('./api/handleSendMatchedUninstrumented.js');
var postMessage = require('./api/postMessage.js');
var i18nextSubliminal = require('i18next-subliminal');

var index = {
  wrap: i18nextSubliminal.wrap,
  unwrap: i18nextSubliminal.unwrap,
  containsHiddenMeta: i18nextSubliminal.containsHiddenMeta,
  PostProcessor: i18nextSubliminal.PostProcessor,
  addLocizeSavedHandler: postMessage.addLocizeSavedHandler,
  locizePlugin: locizePlugin.locizePlugin,
  locizeEditorPlugin: locizePlugin.locizeEditorPlugin,
  setEditorLng: postMessage.setEditorLng,
  startStandalone: startStandalone.startStandalone
};

exports.locizeEditorPlugin = locizePlugin.locizeEditorPlugin;
exports.locizePlugin = locizePlugin.locizePlugin;
exports.startStandalone = startStandalone.startStandalone;
exports.addLocizeSavedHandler = postMessage.addLocizeSavedHandler;
exports.setEditorLng = postMessage.setEditorLng;
Object.defineProperty(exports, 'PostProcessor', {
  enumerable: true,
  get: function () { return i18nextSubliminal.PostProcessor; }
});
Object.defineProperty(exports, 'containsHiddenMeta', {
  enumerable: true,
  get: function () { return i18nextSubliminal.containsHiddenMeta; }
});
Object.defineProperty(exports, 'unwrap', {
  enumerable: true,
  get: function () { return i18nextSubliminal.unwrap; }
});
Object.defineProperty(exports, 'wrap', {
  enumerable: true,
  get: function () { return i18nextSubliminal.wrap; }
});
exports["default"] = index;
